### DISCUSSION TITLE
`nb::enum_<...>.def_prop_ro(...)` after commit `d25bde8f`

### DISCUSSION THREAD

**Discussion by francesco-ballarin** [wjakob/nanobind#556] [Q&A] [answered]
Hi,
in `basix`, one of the components of FEniCS, we have a few enums with a readonly property, e.g.
```
  nb::enum_<lattice::type>(m, "LatticeType")
      .value("equispaced", lattice::type::equispaced)
      .value("gll", lattice::type::gll)
      .value("chebyshev", lattice::type::chebyshev)
      .value("gl", lattice::type::gl)
      .def_prop_ro("name",
                   [](nb::object obj) { return nb::getattr(obj, "__name__"); });
```
at https://github.com/FEniCS/basix/blob/baff38a39049369b382752669b1a4667434b03a0/python/wrapper.cpp#L423-L428

Compiling with current `nanobind` `main` branch, we get errors like
```
      /tmp/basix-src/python/wrapper.cpp:428:8: error: ‘class nanobind::enum_<basix::lattice::type>’ has no member named ‘def_prop_ro’
        428 |       .def_prop_ro("name",
            |        ^~~~~~~~~~~
```
which, by looking at recent git history, may be related to the change in https://github.com/wjakob/nanobind/commit/d25bde8fabaa90dd08530f3748622ed7bee474e8#diff-92b184679f711c1aabb5bcd64d973335fac346586e7027ffa3d12b5f5ccb765c

Can you suggest us how to obtain something equivalent to `nb::enum_<...>.def_prop_ro(...)` in nanobind 2.x?

Thanks,
Francesco

**Comment by wjakob** [comment]
As a workaround, you could assign the enum binding to a variable and then write:

``my_enum.attr("name") = nb::cpp_function(..)`` where you copy-paste your previous lambda into the `..` part.

**Reply by francesco-ballarin** [reply]
> `my_enum.attr("name") = nb::cpp_function(..)`

I tried, but it seems to me that it defines a method, rather than a property. Is there any way I can wrap the `nb::cpp_function` into a property?

FYI, the modified snippet above reads

```
void attach_name(nb::module_& m, std::string enum_type)
{
  m.attr(enum_type.c_str()).attr("name") = nb::cpp_function(
    [](nb::object obj) { return nb::getattr(obj, "__name__"); });
}

NB_MODULE(..., m)
{
  nb::enum_<lattice::type>(m, "LatticeType")
      .value("equispaced", lattice::type::equispaced)
      .value("gll", lattice::type::gll)
      .value("chebyshev", lattice::type::chebyshev)
      .value("gl", lattice::type::gl);
  attach_name(m, "LatticeType");
}
```

**Reply by francesco-ballarin** [reply]
Hi @wjakob,

> I tried, but it seems to me that it defines a method, rather than a property. Is there any way I can wrap the `nb::cpp_function` into a property?

do you have any suggestion for us on how to do this?

If you feel that this is not the right place for this discussion (e.g., this should be reported as an issue) please let me know.

Thanks,
Francesco & the FEniCS team

**Comment by francesco-ballarin** [comment]
Thanks, I will try that.

> As a workaround

Do you expect this workaround to be the way to go in future, or do you expect `def_prop_ro` to come back at some point before 2.0 is released?
(No big deal either way, just to understand how to phrase the change to the other developers in the downstream project when creating the associated PR).

**Reply by wjakob** [reply, marked as answer]
See 2889d773e18af6ad94838f9dc2c20efc079766f4

**Reply by wjakob** [reply]
Please let me know if that fixes the issues.

**Reply by francesco-ballarin** [reply]
Thanks @wjakob , I can confirm that the commit does fix the issue for us!
